pythia/modules/metrics.py

```diff
@@ -198,6 +199,86 @@ def calculate(self, sample_list, model_output, *args, **kwargs):
         return value
 
 
+@registry.register_metric("caption_accuracy")
+class CaptionAccuracy(BaseMetric):
+    """Metric for calculating caption accuracy. Currently we are calculating BLEU4 Score.
+
+    **Key:** ``caption_accuracy``
+    """
+
+    def __init__(self):
+        super().__init__("caption_accuracy")
+        text_processor = registry.get("coco_text_processor")
+        self.vocab = text_processor.vocab
+
+    def calculate(self, sample_list, model_output, *args, **kwargs):
+        """Calculate accuracy and return it back.
+
+        Args:
+            sample_list (SampleList): SampleList provided by DataLoader for
+                                current iteration
+            model_output (Dict): Dict returned by model.
+
+        Returns:
+            torch.FloatTensor: bleu4 score.
+
+        """
+        # Create reference and hypotheses captions.
+        references = list()
+        hypotheses = list()
+
+        # References
+        targets = sample_list.answers
+        for j, p in enumerate(targets):
+            img_caps = targets[j].tolist()
+            img_captions = list(
+                map(
+                    lambda c: [
+                        self.vocab.get_itos()[w]
+                        for w in c
+                        if w
+                        not in {
+                            self.vocab.SOS_INDEX,
+                            self.vocab.EOS_INDEX,
+                            self.vocab.PAD_INDEX,
+                        }
+                    ],
+                    img_caps,
+                )
+            )
+            img_captions = [" ".join(c) for c in img_captions]
+            references.append(img_captions)
+
+        # Hypotheses
+        scores = model_output["scores"]
+        _, scores = torch.max(scores, dim=2)
+        scores = scores.tolist()
+        predictions = list()
+        for j, p in enumerate(scores):
+            for idx, v in enumerate(scores[j]):
```

Comment:
NIT: Better variables names for readability.

Reply:
I will move this logic to a caption processor.